STM32H7: Set voltage scaling before PLL configuration to fix early usage fault

marwaiehm-st opened the pull request:
set voltage scaling before PLL configuration to prevent early usage fault on STM32H7:

Move the call to prepare_regulator_voltage_scale() before PLL setup in clock_stm32_ll_h7.c.

This ensures the voltage regulator is configured to the appropriate voltage scale before increasing
the system clock frequency via PLLs. Without this change, the CPU could run at a frequency unsupported
by the current voltage level, causing z_arm_usage_fault() during early boot on boards like NUCLEO_H723ZG.

This fix resolves stability issues observed in high-frequency workloads such as the zperf sample.

rlubos reviewed (commented):
I can cofirm that I no longer observe issue https://github.com/zephyrproject-rtos/zephyr/issues/88974 on this PR.

marwaiehm-st commented:
> Moreover, the [nucleo_h723zg](https://github.com/zephyrproject-rtos/zephyr/blob/4ec435165753bf353b35ae700c3ed5a1421eb46b/boards/st/nucleo_h723zg/nucleo_h723zg.dts#L117C30-L117C31) is configured to run at 550MHz, which requires `VSO0` (_by docs, by cube_).
> 
> Unfortunately, I don't have that particular board. @rlubos, could you try with the following changes while using `VSO0`:
> 
> ```diff
> --- a/drivers/clock_control/clock_stm32_ll_h7.c
> +++ b/drivers/clock_control/clock_stm32_ll_h7.c
> @@ -1015,15 +1015,15 @@ int stm32_clock_control_init(const struct device *dev)
>  	/* Set up individual enabled clocks */
>  	set_up_fixed_clock_sources();
>  
> +	/* Configure Voltage scale to comply with the desired system frequency */
> +	prepare_regulator_voltage_scale();
> +
>  	/* Set up PLLs */
>  	r = set_up_plls();
>  	if (r < 0) {
>  		return r;
>  	}
>  
> -	/* Configure Voltage scale to comply with the desired system frequency */
> -	prepare_regulator_voltage_scale();
> -
>  	/* Current hclk value */
>  	old_hclk_freq = get_hclk_frequency();
>  	/* AHB is HCLK clock to configure */
> ```
> 
> To configure PLLs after the voltage scale is set.

Hi @GeorgeCGV, 
I tested this modification on the Nucleo_H723ZG board, and it resolved the issue.

marwaiehm-st committed:
drivers: clock_control: set voltage scaling before PLL

Move the call to prepare_regulator_voltage_scale()
before PLL setup in clock_stm32_ll_h7.c.

This ensures the voltage regulator is configured
to the appropriate voltage scale before increasing
the system clock frequency via PLLs.
Without this change, the system configuration
may be out of spec.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

etienne-lms reviewed (approved):
Suggestion in commit message:
```patch
-Without this change, the CPU could run
-at a frequency unsupported
+Without this change, the CPU could run at an unsupported frequency.
```
Or maybe better:
```patch
+Without this change, the system configuration may be out of spec.

GeorgeCGV commented:
Ideally, this might need a backport

etienne-lms commented:
I don't think this fix needs to be backported. According to Issue https://github.com/zephyrproject-rtos/zephyr/issues/88974, the issue is a regression induced by commit https://github.com/zephyrproject-rtos/zephyr/commit/b2b6b9be7ee85c24f620f9055008891f020ab8d4, merged after release tag 4.1.0. Looking at branch *v4.1-branch* latest commit (https://github.com/zephyrproject-rtos/zephyr/commit/801a18aae1bdff334bf450b7d57e718cd4a1d2b0),  the issue is not present on that branch.
